File: AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.java
/****************************************************************************************
 * Copyright (c) 2018 Mike Hardy <mike@mikehardy.net>                                   *
 *                                                                                      *
 * This program is free software; you can redistribute it and/or modify it under        *
 * the terms of the GNU General Public License as published by the Free Software        *
 * Foundation; either version 3 of the License, or (at your option) any later           *
 * version.                                                                             *
 *                                                                                      *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY      *
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A      *
 * PARTICULAR PURPOSE. See the GNU General Public License for more details.             *
 *                                                                                      *
 * You should have received a copy of the GNU General Public License along with         *
 * this program.  If not, see <http://www.gnu.org/licenses/>.                           *
 ****************************************************************************************/

package com.ichi2.utils;

import android.app.Activity;
import android.content.ContentResolver;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.os.Message;
import android.provider.OpenableColumns;
import android.text.TextUtils;
import android.webkit.MimeTypeMap;

import com.afollestad.materialdialogs.MaterialDialog;
import com.ichi2.anki.AnkiActivity;
import com.ichi2.anki.AnkiDroidApp;
import com.ichi2.anki.R;
import com.ichi2.anki.dialogs.DialogHandler;
import com.ichi2.compat.CompatHelper;

import org.apache.commons.compress.archivers.zip.ZipFile;
import org.jetbrains.annotations.Contract;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.util.Locale;
import java.util.zip.ZipEntry;
import java.util.zip.ZipException;
import java.util.zip.ZipInputStream;

import androidx.annotation.CheckResult;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import timber.log.Timber;

public class ImportUtils {

    /* A filename should be shortened if over this threshold */
    private static final int fileNameShorteningThreshold = 100;


    /**
     * This code is used in multiple places to handle package imports
     *
     * @param context for use in resource resolution and path finding
     * @param intent contains the file to import
     * @return null if successful, otherwise error message
     */
    @NonNull
    public static ImportResult handleFileImport(Context context, Intent intent) {
        return new FileImporter().handleFileImport(context, intent);
    }

    public static void showImportUnsuccessfulDialog(Activity activity, String errorMessage, boolean exitActivity) {
        new FileImporter().showImportUnsuccessfulDialog(activity, errorMessage, exitActivity);
    }

    public static boolean isCollectionPackage(String filename) {
        return filename != null && (filename.toLowerCase(Locale.ROOT).endsWith(".colpkg") || "collection.apkg".equals(filename));
    }

    /** @return Whether the file is either a deck, or a collection package */
    @Contract("null -> false")
    public static boolean isValidPackageName(@Nullable String filename) {
        return FileImporter.isDeckPackage(filename) || isCollectionPackage(filename);
    }

    /**
     * Whether importUtils can handle the given intent
     * Caused by #6312 - A launcher was sending ACTION_VIEW instead of ACTION_MAIN
     */
    public static boolean isInvalidViewIntent(@NonNull Intent intent) {
        return intent.getData() == null && intent.getClipData() == null;
    }
    
    public static boolean isFileAValidDeck(String fileName) {
        return FileImporter.hasExtension(fileName, "apkg") || FileImporter.hasExtension(fileName, "colpkg");
    }


    @SuppressWarnings("WeakerAccess")
    protected static class FileImporter {
        /**
         * This code is used in multiple places to handle package imports
         *
         * @param context for use in resource resolution and path finding
         * @param intent contains the file to import
         * @return null if successful, otherwise error message
         */
        @NonNull
        public ImportResult handleFileImport(Context context, Intent intent) {
            // This intent is used for opening apkg package files
            // We want to go immediately to DeckPicker, clearing any history in the process
            Timber.i("IntentHandler/ User requested to view a file");
            String extras = intent.getExtras() == null ? "none" : TextUtils.join(", ", intent.getExtras().keySet());
            Timber.i("Intent: %s. Data: %s", intent, extras);

            try {
                return handleFileImportInternal(context, intent);
            } catch (Exception e) {
                AnkiDroidApp.sendExceptionReport(e, "handleFileImport");
                Timber.e(e, "failed to handle import intent");
                return ImportResult.fromErrorString(context.getString(R.string.import_error_exception, e.getLocalizedMessage()));
            }
        }

        //Added to remove exception handlers
        @NonNull
        private ImportResult handleFileImportInternal(Context context, Intent intent) {
            if (intent.getData() == null) {
                Timber.i("No intent data. Attempting to read clip data.");
                if (intent.getClipData() == null
                        || intent.getClipData().getItemCount() == 0) {
                    return ImportResult.fromErrorString(context.getString(R.string.import_error_unhandled_request));
                }
                Uri clipUri = intent.getClipData().getItemAt(0).getUri();
                return handleContentProviderFile(context, intent, clipUri);
            }

            // If Uri is of scheme which is supported by ContentResolver, read the contents
            String intentUriScheme = intent.getData().getScheme();
            if (intentUriScheme.equals(ContentResolver.SCHEME_CONTENT) || intentUriScheme.equals(ContentResolver.SCHEME_FILE)
                    || intentUriScheme.equals(ContentResolver.SCHEME_ANDROID_RESOURCE)) {
                Timber.i("Attempting to read content from intent.");
                return handleContentProviderFile(context, intent, intent.getData());
            } else {
                return ImportResult.fromErrorString(context.getResources().getString(R.string.import_error_unhandled_scheme, intent.getData()));
            }
        }


        @NonNull
        private ImportResult handleContentProviderFile(Context context, Intent intent, Uri data) {
            //Note: intent.getData() can be null. Use data instead.

            // Get the original filename from the content provider URI
            String filename = getFileNameFromContentProvider(context, data);

            // Hack to fix bug where ContentResolver not returning filename correctly
            if (filename == null) {
                if (intent.getType() != null && ("application/apkg".equals(intent.getType()) || hasValidZipFile(context, data))) {
                    // Set a dummy filename if MIME type provided or is a valid zip file
                    filename = "unknown_filename.apkg";
                    Timber.w("Could not retrieve filename from ContentProvider, but was valid zip file so we try to continue");
                } else {
                    Timber.e("Could not retrieve filename from ContentProvider or read content as ZipFile");
                    AnkiDroidApp.sendExceptionReport(new RuntimeException("Could not import apkg from ContentProvider"), "IntentHandler.java", "apkg import failed");
                    return ImportResult.fromErrorString(AnkiDroidApp.getAppResources().getString(R.string.import_error_content_provider, AnkiDroidApp.getManualUrl() + "#importing"));
                }
            }

            if (!isValidPackageName(filename)) {
                if (isAnkiDatabase(filename)) {
                    //.anki2 files aren't supported by Anki Desktop, we should eventually support them, because we can
                    //but for now, show a "nice" error.
                    return ImportResult.fromErrorString(context.getResources().getString(R.string.import_error_load_imported_database));
                } else {
                    // Don't import if file doesn't have an Anki package extension
                    return ImportResult.fromErrorString(context.getResources().getString(R.string.import_error_not_apkg_extension, filename));
                }
            } else {
                // Copy to temporary file
                filename = ensureValidLength(filename);
                String tempOutDir = Uri.fromFile(new File(context.getCacheDir(), filename)).getEncodedPath();
                String errorMessage = copyFileToCache(context, data, tempOutDir) ? null : context.getString(R.string.import_error_copy_file_to_cache);
                // Show import dialog
                if (errorMessage != null) {
                    AnkiDroidApp.sendExceptionReport(new RuntimeException("Error importing apkg file"), "IntentHandler.java", "apkg import failed");
                    return ImportResult.fromErrorString(errorMessage);
                }

                ImportResult validateZipResult = validateZipFile(context, tempOutDir);
                if (validateZipResult != null) {
                    //noinspection ResultOfMethodCallIgnored
                    new File(tempOutDir).delete();
                    return validateZipResult;
                }

                sendShowImportFileDialogMsg(tempOutDir);
                return ImportResult.fromSuccess();
            }
        }

        @Nullable
        protected ImportResult validateZipFile(Context ctx, String filePath) {
            File file = new File(filePath);
            ZipFile zf = null;
            try {
                zf = new ZipFile(file);
            } catch (Exception e) {
                Timber.w(e, "Failed to validate zip");
                return ImportResult.fromInvalidZip(ctx, file, e);
            } finally {
                if (zf != null) {
                    try {
                        zf.close();
                    } catch (IOException e) {
                        Timber.w(e, "Failed to close zip");
                    }
                }
            }
            return null;
        }


        private boolean isAnkiDatabase(String filename) {
            return filename != null && hasExtension(filename, "anki2");
        }


        private String ensureValidLength(String fileName) {
            //#6137 - filenames can be too long when URLEncoded
            try {
                String encoded = URLEncoder.encode(fileName, "UTF-8");

                if (encoded.length() <= fileNameShorteningThreshold) {
                    Timber.d("No filename truncation necessary");
                    return fileName;
                } else {
                    Timber.d("Filename was longer than %d, shortening", fileNameShorteningThreshold);
                    //take 90 instead of 100 so we don't get the extension
                    int substringLength = fileNameShorteningThreshold - 10;
                    String shortenedFileName = encoded.substring(0, substringLength) + "..." + getExtension(fileName);
                    Timber.d("Shortened filename '%s' to '%s'", fileName, shortenedFileName);
                    //if we don't decode, % is double-encoded
                    return URLDecoder.decode(shortenedFileName, "UTF-8");
                }
            } catch (Exception e) {
                Timber.w(e, "Failed to shorten file: %s", fileName);
                return fileName;
            }
        }

        @CheckResult
        private String getExtension(String fileName) {
            Uri file = Uri.fromFile(new File(fileName));
            return MimeTypeMap.getFileExtensionFromUrl(file.toString());
        }


        @Nullable
        protected String getFileNameFromContentProvider(Context context, Uri data) {
            String filename = null;
            try (Cursor cursor = context.getContentResolver().query(data, new String[] {OpenableColumns.DISPLAY_NAME}, null, null, null)) {
                if (cursor != null && cursor.moveToFirst()) {
                    filename = cursor.getString(0);
                    Timber.d("handleFileImport() Importing from content provider: %s", filename);
                }
            }
            return filename;
        }


        public void showImportUnsuccessfulDialog(Activity activity, String errorMessage, boolean exitActivity) {
            Timber.e("showImportUnsuccessfulDialog() message %s", errorMessage);
            String title = activity.getResources().getString(R.string.import_log_no_apkg);
            new MaterialDialog.Builder(activity)
                    .title(title)
                    .content(errorMessage)
                    .positiveText(activity.getResources().getString(R.string.dialog_ok))
                    .onPositive((dialog, which) -> {
                        if (exitActivity) {
                            AnkiActivity.finishActivityWithFade(activity);
                        }
                    })
                    .build().show();
        }


        /**
         * Send a Message to AnkiDroidApp so that the DialogMessageHandler shows the Import apkg dialog.
         * @param path path to apkg file which will be imported
         */
        private static void sendShowImportFileDialogMsg(String path) {
            // Get the filename from the path
            File f = new File(path);
            String filename = f.getName();

            // Create a new message for DialogHandler so that we see the appropriate import dialog in DeckPicker
            Message handlerMessage = Message.obtain();
            Bundle msgData = new Bundle();
            msgData.putString("importPath", path);
            handlerMessage.setData(msgData);
            if (isCollectionPackage(filename)) {
                // Show confirmation dialog asking to confirm import with replace when file called "collection.apkg"
                handlerMessage.what = DialogHandler.MSG_SHOW_COLLECTION_IMPORT_REPLACE_DIALOG;
            } else {
                // Otherwise show confirmation dialog asking to confirm import with add
                handlerMessage.what = DialogHandler.MSG_SHOW_COLLECTION_IMPORT_ADD_DIALOG;
            }
            // Store the message in AnkiDroidApp message holder, which is loaded later in AnkiActivity.onResume
            DialogHandler.storeMessage(handlerMessage);
        }

        private static boolean isDeckPackage(String filename) {
            return filename != null && filename.toLowerCase(Locale.ROOT).endsWith(".apkg") && !"collection.apkg".equals(filename);
        }


        public static boolean hasExtension(@NonNull String filename, String extension) {
            String[] fileParts = filename.split("\\.");
            if (fileParts.length < 2) {
                return false;
            }
            String extensionSegment = fileParts[fileParts.length - 1];
            //either "apkg", or "apkg (1)".
            // COULD_BE_BETTE: accepts .apkgaa"
            return extensionSegment.toLowerCase(Locale.ROOT).startsWith(extension);
        }


        /**
         * Check if the InputStream is to a valid non-empty zip file
         * @param data uri from which to get input stream
         * @return whether or not valid zip file
         */
        private static boolean hasValidZipFile(Context context, Uri data) {
            // Get an input stream to the data in ContentProvider
            InputStream in = null;
            try {
                in = context.getContentResolver().openInputStream(data);
            } catch (FileNotFoundException e) {
                Timber.e(e, "Could not open input stream to intent data");
            }
            // Make sure it's not null
            if (in == null) {
                Timber.e("Could not open input stream to intent data");
                return false;
            }
            // Open zip input stream
            ZipInputStream zis = new ZipInputStream(in);
            boolean ok = false;
            try {
                try {
                    ZipEntry ze = zis.getNextEntry();
                    if (ze != null) {
                        // set ok flag to true if there are any valid entries in the zip file
                        ok = true;
                    }
                } catch (Exception e) {
                    // don't set ok flag
                    Timber.d(e, "Error checking if provided file has a zip entry");
                }
            } finally {
                // close the input streams
                try {
                    zis.close();
                    in.close();
                } catch (Exception e) {
                    Timber.d(e, "Error closing the InputStream");
                }
            }
            return ok;
        }


        /**
         * Copy the data from the intent to a temporary file
         * @param data intent from which to get input stream
         * @param tempPath temporary path to store the cached file
         * @return whether or not copy was successful
         */
        protected boolean copyFileToCache(Context context, Uri data, String tempPath) {
            // Get an input stream to the data in ContentProvider
            InputStream in;
            try {
                in = context.getContentResolver().openInputStream(data);
            } catch (FileNotFoundException e) {
                Timber.e(e, "Could not open input stream to intent data");
                return false;
            }
            // Check non-null
            if (in == null) {
                return false;
            }

            try {
                CompatHelper.getCompat().copyFile(in, tempPath);
            } catch (IOException e) {
                Timber.e(e, "Could not copy file to %s", tempPath);
                return false;
            } finally {
                try {
                    in.close();
                } catch (IOException e) {
                    Timber.e(e, "Error closing input stream");
                }
            }
            return true;
        }
    }

    public static class ImportResult {
        private final String mMessage;


        public ImportResult(String message) {
            this.mMessage = message;
        }

        public static ImportResult fromErrorString(String message) {
            return new ImportResult(message);
        }

        public static ImportResult fromSuccess() {
            return new ImportResult(null);
        }


        public static ImportResult fromInvalidZip(Context ctx, File file, Exception e) {
            return fromErrorString(getInvalidZipException(ctx, file, e));
        }


        private static String getInvalidZipException(Context ctx, @SuppressWarnings( {"unused", "RedundantSuppression"}) File file, Exception e) {
            // This occurs when there is random corruption in a zip file
            if (e instanceof IOException && "central directory is empty, can't expand corrupt archive.".equals(e.getMessage())) {
                return ctx.getString(R.string.import_error_corrupt_zip, e.getLocalizedMessage());
            }
            // 7050 - this occurs when a file is truncated at the end (partial download/corrupt).
            if (e instanceof ZipException && "archive is not a ZIP archive".equals(e.getMessage())) {
                return ctx.getString(R.string.import_error_corrupt_zip, e.getLocalizedMessage());
            }

            // If we don't have a good string, send a silent exception that we can better handle this in the future
            AnkiDroidApp.sendExceptionReport(e, "Import - invalid zip", "improve UI message here", true);
            return ctx.getString(R.string.import_log_failed_unzip, e.getLocalizedMessage());
        }


        public boolean isSuccess() {
            return mMessage == null;
        }

        public String getHumanReadableMessage() {
            return mMessage;
        }
    }
}

File: AnkiDroid/src/main/java/com/ichi2/utils/ImportUtils.kt
/****************************************************************************************
 * Copyright (c) 2018 Mike Hardy <mike@mikehardy.net>                                   *
 *                                                                                      *
 * This program is free software; you can redistribute it and/or modify it under        *
 * the terms of the GNU General Public License as published by the Free Software        *
 * Foundation; either version 3 of the License, or (at your option) any later           *
 * version.                                                                             *
 *                                                                                      *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY      *
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A      *
 * PARTICULAR PURPOSE. See the GNU General Public License for more details.             *
 *                                                                                      *
 * You should have received a copy of the GNU General Public License along with         *
 * this program.  If not, see <http://www.gnu.org/licenses/>.                           *
 ****************************************************************************************/

package com.ichi2.utils

import android.app.Activity
import android.content.ContentResolver
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Message
import android.provider.OpenableColumns
import androidx.annotation.CheckResult
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.view.ContextThemeWrapper
import androidx.core.os.bundleOf
import com.ichi2.anki.*
import com.ichi2.anki.dialogs.DialogHandler
import com.ichi2.anki.dialogs.DialogHandlerMessage
import com.ichi2.anki.dialogs.ImportDialog
import com.ichi2.compat.CompatHelper
import org.jetbrains.annotations.Contract
import timber.log.Timber
import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream
import java.net.URLDecoder
import java.net.URLEncoder
import java.util.*

object ImportUtils {
    /* A filename should be shortened if over this threshold */
    private const val fileNameShorteningThreshold = 100

    /**
     * This code is used in multiple places to handle package imports
     *
     * @param context for use in resource resolution and path finding
     * @param intent contains the file to import
     * @return null if successful, otherwise error message
     */
    fun handleFileImport(context: Context, intent: Intent): ImportResult {
        return FileImporter().handleFileImport(context, intent)
    }

    /**
     * Makes a cached copy of the file selected on [intent] and returns its path
     */
    fun getFileCachedCopy(context: Context, intent: Intent): String? {
        return FileImporter().getFileCachedCopy(context, intent)
    }

    fun getFileCachedCopy(context: Context, uri: Uri): String? {
        return FileImporter().getFileCachedCopy(context, uri)
    }

    fun showImportUnsuccessfulDialog(activity: Activity, errorMessage: String?, exitActivity: Boolean) {
        FileImporter().showImportUnsuccessfulDialog(activity, errorMessage, exitActivity)
    }

    fun isCollectionPackage(filename: String?): Boolean {
        return filename != null && (filename.lowercase(Locale.ROOT).endsWith(".colpkg") || "collection.apkg" == filename)
    }

    /** @return Whether the file is either a deck, or a collection package */
    @Contract("null -> false")
    fun isValidPackageName(filename: String?): Boolean {
        return FileImporter.isDeckPackage(filename) || isCollectionPackage(filename)
    }

    /**
     * Whether importUtils can handle the given intent
     * Caused by #6312 - A launcher was sending ACTION_VIEW instead of ACTION_MAIN
     */
    fun isInvalidViewIntent(intent: Intent): Boolean {
        return intent.data == null && intent.clipData == null
    }

    fun isFileAValidDeck(fileName: String): Boolean {
        return FileImporter.hasExtension(fileName, "apkg") || FileImporter.hasExtension(fileName, "colpkg")
    }

    @SuppressWarnings("WeakerAccess")
    open class FileImporter {
        /**
         * This code is used in multiple places to handle package imports
         *
         * @param context for use in resource resolution and path finding
         * @param intent contains the file to import
         * @return null if successful, otherwise error message
         */
        fun handleFileImport(context: Context, intent: Intent): ImportResult {
            // This intent is used for opening apkg package files
            // We want to go immediately to DeckPicker, clearing any history in the process
            Timber.i("IntentHandler/ User requested to view a file")
            val extras = if (intent.extras == null) "none" else intent.extras!!.keySet().joinToString(", ")
            Timber.i("Intent: %s. Data: %s", intent, extras)
            return try {
                handleFileImportInternal(context, intent)
            } catch (e: Exception) {
                CrashReportService.sendExceptionReport(e, "handleFileImport")
                Timber.e(e, "failed to handle import intent")
                ImportResult.fromErrorString(context.getString(R.string.import_error_handle_exception, e.localizedMessage))
            }
        }

        private fun handleFileImportInternal(context: Context, intent: Intent): ImportResult {
            val importPathUri = getDataUri(intent)
            return if (importPathUri != null) {
                handleContentProviderFile(context, intent, importPathUri)
            } else {
                ImportResult.fromErrorString(context.getString(R.string.import_error_handle_exception))
            }
        }

        fun getFileCachedCopy(context: Context, uri: Uri): String? {
            val filename = ensureValidLength(getFileNameFromContentProvider(context, uri) ?: return null)
            val tempPath = Uri.fromFile(File(context.cacheDir, filename)).encodedPath!!
            return if (copyFileToCache(context, uri, tempPath)) {
                tempPath
            } else {
                null
            }
        }

        /**
         * Makes a cached copy of the file selected on [intent] and returns its path
         */
        fun getFileCachedCopy(context: Context, intent: Intent): String? {
            val uri = getDataUri(intent) ?: return null
            return getFileCachedCopy(context, uri)
        }

        private fun handleContentProviderFile(
            context: Context,
            intent: Intent,
            importPathUri: Uri
        ): ImportResult {
            // Note: intent.getData() can be null. Use data instead.
            if (!isValidImportType(context, importPathUri)) {
                return ImportResult.fromErrorString(context.getString(R.string.import_log_no_apkg))
            }
            // Get the original filename from the content provider URI
            var filename = getFileNameFromContentProvider(context, importPathUri)
            // Hack to fix bug where ContentResolver not returning filename correctly
            if (filename == null) {
                if (intent.type == "application/apkg" || intent.type == "application/zip") {
                    // Set a dummy filename if MIME type provided or is a valid zip file
                    filename = "unknown_filename.apkg"
                    Timber.w("Could not retrieve filename from ContentProvider, but was valid zip file so we try to continue")
                } else {
                    Timber.e("Could not retrieve filename from ContentProvider")
                    CrashReportService.sendExceptionReport(
                        RuntimeException("Could not import apkg from ContentProvider"),
                        "IntentHandler.java",
                        "apkg import failed; mime type ${intent.type}"
                    )
                    return ImportResult.fromErrorString(AnkiDroidApp.appResources.getString(R.string.import_error_content_provider, AnkiDroidApp.manualUrl + "#importing"))
                }
            }
            val tempOutDir: String
            if (!isValidPackageName(filename)) {
                return if (isAnkiDatabase(filename)) {
                    // .anki2 files aren't supported by Anki Desktop, we should eventually support them, because we can
                    // but for now, show a "nice" error.
                    ImportResult.fromErrorString(context.resources.getString(R.string.import_error_load_imported_database))
                } else {
                    // Don't import if file doesn't have an Anki package extension
                    ImportResult.fromErrorString(context.resources.getString(R.string.import_error_not_apkg_extension, filename))
                }
            } else {
                // Copy to temporary file
                filename = ensureValidLength(filename)
                tempOutDir = Uri.fromFile(File(context.cacheDir, filename)).encodedPath!!
                val errorMessage = if (copyFileToCache(context, importPathUri, tempOutDir)) {
                    null
                } else {
                    context.getString(R.string.import_error_copy_to_cache)
                }
                // Show import dialog
                if (errorMessage != null) {
                    CrashReportService.sendExceptionReport(
                        RuntimeException("Error importing apkg file"),
                        "IntentHandler.java",
                        "apkg import failed"
                    )
                    return ImportResult.fromErrorString(errorMessage)
                }
            }
            sendShowImportFileDialogMsg(tempOutDir)
            return ImportResult.fromSuccess()
        }

        private fun isValidImportType(context: Context, importPathUri: Uri): Boolean {
            val fileName = getFileNameFromContentProvider(context, importPathUri)
            return when {
                isDeckPackage(fileName) -> true
                isCollectionPackage(fileName) -> true
                else -> false
            }
        }

        private fun isAnkiDatabase(filename: String?): Boolean {
            return filename != null && hasExtension(filename, "anki2")
        }

        private fun ensureValidLength(fileName: String): String {
            // #6137 - filenames can be too long when URLEncoded
            return try {
                val encoded = URLEncoder.encode(fileName, "UTF-8")
                if (encoded.length <= fileNameShorteningThreshold) {
                    Timber.d("No filename truncation necessary")
                    fileName
                } else {
                    Timber.d("Filename was longer than %d, shortening", fileNameShorteningThreshold)
                    // take 90 instead of 100 so we don't get the extension
                    val substringLength = fileNameShorteningThreshold - 10
                    val shortenedFileName = encoded.substring(0, substringLength) + "..." + getExtension(fileName)
                    Timber.d("Shortened filename '%s' to '%s'", fileName, shortenedFileName)
                    // if we don't decode, % is double-encoded
                    URLDecoder.decode(shortenedFileName, "UTF-8")
                }
            } catch (e: Exception) {
                Timber.w(e, "Failed to shorten file: %s", fileName)
                fileName
            }
        }

        @CheckResult
        private fun getExtension(fileName: String): String {
            val file = Uri.fromFile(File(fileName))
            return AssetHelper.getFileExtensionFromFilePath(file.toString())
        }

        protected open fun getFileNameFromContentProvider(context: Context, data: Uri): String? {
            var filename: String? = null
            context.contentResolver.query(data, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null).use { cursor ->
                if (cursor != null && cursor.moveToFirst()) {
                    filename = cursor.getString(0)
                    Timber.d("handleFileImport() Importing from content provider: %s", filename)
                }
            }
            return filename
        }

        fun showImportUnsuccessfulDialog(activity: Activity, errorMessage: String?, exitActivity: Boolean) {
            Timber.e("showImportUnsuccessfulDialog() message %s", errorMessage)
            val title = activity.resources.getString(R.string.import_title_error)
            AlertDialog.Builder(ContextThemeWrapper(activity, R.style.AlertDialogStyle)).show {
                title(text = title)
                message(text = errorMessage!!)
                setCancelable(false)
                positiveButton(R.string.dialog_ok) {
                    if (exitActivity) {
                        activity.finish()
                    }
                }
            }
        }

        /**
         * Copy the data from the intent to a temporary file
         * @param data intent from which to get input stream
         * @param tempPath temporary path to store the cached file
         * @return whether or not copy was successful
         */
        protected open fun copyFileToCache(context: Context, data: Uri?, tempPath: String): Boolean {
            // Get an input stream to the data in ContentProvider
            val inputStream: InputStream = try {
                context.contentResolver.openInputStream(data!!)
            } catch (e: FileNotFoundException) {
                Timber.e(e, "Could not open input stream to intent data")
                return false
            } ?: return false
            // Check non-null
            try {
                CompatHelper.compat.copyFile(inputStream, tempPath)
            } catch (e: IOException) {
                Timber.e(e, "Could not copy file to %s", tempPath)
                return false
            } finally {
                try {
                    inputStream.close()
                } catch (e: IOException) {
                    Timber.e(e, "Error closing input stream")
                }
            }
            return true
        }

        companion object {
            fun getDataUri(intent: Intent): Uri? {
                if (intent.data == null) {
                    Timber.i("No intent data. Attempting to read clip data.")
                    if (intent.clipData == null || intent.clipData!!.itemCount == 0) {
                        return null
                    }
                    return intent.clipData?.getItemAt(0)?.uri
                }
                // If Uri is of scheme which is supported by ContentResolver, read the contents
                val intentUriScheme = intent.data!!.scheme
                return if (intentUriScheme == ContentResolver.SCHEME_CONTENT || intentUriScheme == ContentResolver.SCHEME_FILE || intentUriScheme == ContentResolver.SCHEME_ANDROID_RESOURCE) {
                    Timber.i("Attempting to read data from intent.")
                    intent.data
                } else {
                    null
                }
            }

            /**
             * Send a Message to AnkiDroidApp so that the DialogMessageHandler shows the Import apkg dialog.
             * @param importPath path of to apkg file which will be imported
             */
            private fun sendShowImportFileDialogMsg(importPath: String) {
                // Get the filename from the path
                val filename = File(importPath).name

                val dialogMessage = if (isCollectionPackage(filename)) {
                    CollectionImportReplace(importPath)
                } else {
                    CollectionImportAdd(importPath)
                }
                // Store the message in AnkiDroidApp message holder, which is loaded later in AnkiActivity.onResume
                DialogHandler.storeMessage(dialogMessage.toMessage())
            }

            internal fun isDeckPackage(filename: String?): Boolean {
                return filename != null && filename.lowercase(Locale.ROOT).endsWith(".apkg") && "collection.apkg" != filename
            }

            fun hasExtension(filename: String, extension: String?): Boolean {
                val fileParts = filename.split("\\.".toRegex()).toTypedArray()
                if (fileParts.size < 2) {
                    return false
                }
                val extensionSegment = fileParts[fileParts.size - 1]
                // either "apkg", or "apkg (1)".
                // COULD_BE_BETTE: accepts .apkgaa"
                return extensionSegment.lowercase(Locale.ROOT).startsWith(extension!!)
            }
        }
    }

    class ImportResult(val humanReadableMessage: String?) {
        val isSuccess: Boolean
            get() = humanReadableMessage == null

        companion object {
            fun fromErrorString(message: String?): ImportResult {
                return ImportResult(message)
            }

            fun fromSuccess(): ImportResult {
                return ImportResult(null)
            }
        }
    }

    /** Show confirmation dialog asking to confirm import with replace when file called "collection.apkg" */
    class CollectionImportReplace(private val importPath: String) : DialogHandlerMessage(
        which = WhichDialogHandler.MSG_SHOW_COLLECTION_IMPORT_REPLACE_DIALOG,
        analyticName = "ImportReplaceDialog"
    ) {
        override fun handleAsyncMessage(deckPicker: DeckPicker) {
            // Handle import of collection package APKG
            deckPicker.showImportDialog(ImportDialog.DIALOG_IMPORT_REPLACE_CONFIRM, importPath)
        }

        override fun toMessage(): Message = Message.obtain().apply {
            data = bundleOf("importPath" to importPath)
            what = this@CollectionImportReplace.what
        }

        companion object {
            fun fromMessage(message: Message): CollectionImportReplace =
                CollectionImportReplace(message.data.getString("importPath")!!)
        }
    }

    /** Show confirmation dialog asking to confirm import with add */
    class CollectionImportAdd(private val importPath: String) : DialogHandlerMessage(
        WhichDialogHandler.MSG_SHOW_COLLECTION_IMPORT_ADD_DIALOG,
        "ImportAddDialog"
    ) {
        override fun handleAsyncMessage(deckPicker: DeckPicker) {
            // Handle import of deck package APKG
            deckPicker.showImportDialog(ImportDialog.DIALOG_IMPORT_ADD_CONFIRM, importPath)
        }

        override fun toMessage(): Message = Message.obtain().apply {
            data = bundleOf("importPath" to importPath)
            what = this@CollectionImportAdd.what
        }

        companion object {
            fun fromMessage(message: Message): CollectionImportAdd =
                CollectionImportAdd(message.data.getString("importPath")!!)
        }
    }
}
